### PR TITLE
[X86] Add "Ws" constraint and "p" modifier for symbolic address/label reference

### PR DESCRIPTION
**Pull request by MaskRay**
Printing the raw symbol is useful in inline asm (e.g. getting the C++
mangled name, referencing a symbol in a custom way while ensuring it is
not optimized out even if internal). Similar constraints are available
in other targets (e.g. "S" for aarch64/riscv, "Cs" for m68k).

```
namespace ns { extern int var, a[4]; }
void foo() {
  asm(".pushsection .xxx,\"aw\"; .dc.a %p0; .popsection" :: "Ws"(&ns::var));
  asm(".reloc ., BFD_RELOC_NONE, %p0" :: "Ws"(&ns::a[3]));
}
```

Link: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105576

### CODE REVIEW

File: llvm/test/CodeGen/X86/inline-asm-Ws-constraint.ll [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+; CHECK-NEXT:  # %bb.1: # %label
+; CHECK-NEXT:    #APP
+; CHECK-NEXT:    # .Ltmp0
+; CHECK-NEXT:    #NO_APP
+; CHECK-NEXT:    ret{{[l|q]}}
+entry:
+  br label %label
+
+label:
+  tail call void asm sideeffect "// ${0:p}", "Ws,~{dirflag},~{fpsr},~{flags}"(ptr blockaddress(@test_label, %label))
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/inline-asm-Ws-constraint.ll:32]: Missing a `^` before `Ws`?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -56966,6 +56965,17 @@ void X86TargetLowering::LowerAsmOperandForConstraint(SDValue Op,
...
   }
+  case 'W': {
+    assert(Constraint[1] == 's');
+    if (const auto *GA = dyn_cast<GlobalAddressSDNode>(Op)) {
+      Ops.push_back(DAG.getTargetGlobalAddress(GA->getGlobal(), SDLoc(Op),
+                                               GA->getValueType(0)));
+    } else if (const auto *BA = dyn_cast<BlockAddressSDNode>(Op)) {
+      Ops.push_back(DAG.getTargetBlockAddress(BA->getBlockAddress(),
+                                              BA->getValueType(0)));
+    }
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:56976]: Don't need parentheses here.

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -5336,6 +5336,7 @@ X86:
   operand in a SSE register. If AVX is also enabled, can also be a 256-bit
   vector operand in an AVX register. If AVX-512 is also enabled, can also be a
   512-bit vector operand in an AVX512 register. Otherwise, an error.
+- ``Ws``: A symbolic reference or label reference.
```
Comment (phoebewang) [inline llvm/docs/LangRef.rst:5339]: What's the reason to choose `Ws`?

Comment (MaskRay) [reply llvm/docs/LangRef.rst:5339]: x86 has almost run out of uppercase and lowercase letters... The remaining ones are `z` and `H`.

I tried `z` but one GCC x86 maintainer considers `W` prefix better: https://gcc.gnu.org/pipermail/gcc-patches/2024-January/642596.html

File: llvm/test/CodeGen/X86/inline-asm-Ws-constraint.ll (hunk 1 above)
Comment (MaskRay) [reply llvm/test/CodeGen/X86/inline-asm-Ws-constraint.ll]: Thanks for spotting this. Fixed.

I think constraint validation is primarily performed in Clang Sema and it's easy to crash LLVM with invalid LLVM IR. In this case Ws+0:p happens to work as well.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 2 above)
Comment (MaskRay) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: Removed

File: llvm/docs/LangRef.rst (hunk 3 above)
Comment (phoebewang) [reply llvm/docs/LangRef.rst:5339]: Got it, thanks for the link! Do we support the pointer offset like GCC `&var + 1`?

Comment (MaskRay) [reply llvm/docs/LangRef.rst:5339]: Yes, it's supported (aarch64 "S" supports it but riscv "S" doesn't) I intend to support it for x86 "Ws".

OK, I've figured out how to do it in SelectionDAG... Updated.

File: clang/test/Sema/inline-asm-validate-x86.c [hunk 4]
```diff
@@ -130,3 +130,7 @@ void pr40890(void) {
   __asm__ __volatile__("\n#define BEEF abcd%0\n" : : "n"((int*)0xdeadbeeeeeef));
 #endif
 }
+
+void test_W(int i) {
+  asm("" : : "Wd"(test_W)); // expected-error{{invalid input constraint 'Wd' in asm}}
```
Comment (phoebewang) [inline clang/test/Sema/inline-asm-validate-x86.c:135]: Use `__asm__` for consistency?

Comment (phoebewang) [reply clang/test/Sema/inline-asm-validate-x86.c:135]: And add a positive test?

File: clang/lib/Basic/Targets/X86.cpp [hunk 5]
```diff
@@ -1418,6 +1418,14 @@ bool X86TargetInfo::validateAsmConstraint(
   case 'O':
     Info.setRequiresImmediate(0, 127);
     return true;
+  case 'W':
+    switch (*++Name) {
+    default:
+      return false;
+    case 's':
+      Info.setAllowsRegister();
```
Comment (phoebewang) [inline clang/lib/Basic/Targets/X86.cpp:1426]: Not sure if it matters to a print constraint, should it be `setAllowsMemory`?

**Review by phoebewang — APPROVED** [review]
LGTM with some nits.

File: clang/lib/Basic/Targets/X86.cpp (hunk 5 above)
Comment (MaskRay) [reply clang/lib/Basic/Targets/X86.cpp:1426]: `setAllowsRegister` is somewhat confusing but its use is correct here (also used by aarch64/riscv `S`).

`setAllowsMemory` seems to not add extra checks.

File: clang/test/Sema/inline-asm-validate-x86.c (hunk 4 above)
Comment (MaskRay) [reply clang/test/Sema/inline-asm-validate-x86.c:135]: The codegen tests  `asm("// %p0 %p1 %p2" :: "Ws"(&var), "Ws"(&arr[3]), "Ws"(test_Ws));` (and a struct test I am adding) provide positive tests:)

**Pull request merged**